File: find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) return new int[]{-1, -1};
        int[] result = new int[]{(int)(Math.pow(10,5)) + 1, -1};

        doSearch(nums, target, result, 0, nums.length-1);
        return result[0] == (int)(Math.pow(10,5)) + 1 & result[1] == -1 ? new int[]{-1, -1} : result;
    }

    private void doSearch(int[] nums, int target, int[] result, int left, int right) {
        if (left > right) return;

        int mid = left + (right - left) / 2;
        if (nums[mid] == target) {
            result[0] = Math.min(result[0], mid);
            result[1] = Math.max(result[1], mid);
            doSearch(nums, target, result, left, mid - 1);
            doSearch(nums, target, result, mid + 1, right);
        } else if (nums[mid] < target) {
            doSearch(nums, target, result, mid + 1, right);
        } else {
            doSearch(nums, target, result, left, mid - 1);
        }
    }
}

File: find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        idx_left = self.findFirstIndex(nums, target)
        if idx_left == len(nums) or nums[idx_left] != target:
            return [-1, -1]
        return [idx_left, self.findFirstIndex(nums, target+1) - 1]

    def findFirstIndex(self, nums: List[int], target: int):
        low, high = 0, len(nums)

        while low < high:
            mid = low + (high - low) // 2
            if nums[mid] < target:
                low = mid + 1
            else:
                high = mid
        return low